app/crud/serviceMetric.py

```diff
@@ -1,19 +1,83 @@
 
 from sqlalchemy.orm import Session
-
-from ..models import ServiceMetric
+from ..models import ServiceMetric, scoreCardMetrics
 from ..schemas import ServiceMetricCreate, ServiceMetricUpdate
 from .base import CRUDBase
+from . import CRUDScoreCardMetric ,CRUDMetric
+from app.utils.utility_functions import main_calculate_error, calculate_score
 
 
 class CRUDServiceMetric(CRUDBase[ServiceMetric, ServiceMetricCreate, ServiceMetricUpdate]):
     def __init__(self, db_session: Session):
         super(CRUDServiceMetric, self).__init__(ServiceMetric, db_session)
+        self.scorecardMetrics = CRUDScoreCardMetric(db_session)
+        self.Metric = CRUDMetric(db_session)
 
     def getByScorecardId(self, scorecardId: int) -> list[ServiceMetric]:
         return self.db_session.query(ServiceMetric).filter(ServiceMetric.scorecardId == scorecardId).all()
 
     def getByServiceId(self, serviceId: int) -> list[ServiceMetric]:
         return self.db_session.query(ServiceMetric).filter(ServiceMetric.serviceId == serviceId).all()
+    
+
+    def get_timestamp(self, service_id: int, scorecard_id: int):
+        subquery = self.scorecardMetrics.getMetricByScoreCradId(scorecard_id)
+
+        update_time = self.db_session.query(ServiceMetric.timestamp)\
+            .filter(ServiceMetric.serviceId == service_id, ServiceMetric.metricId.in_(subquery))\
+            .order_by(ServiceMetric.timestamp.desc())\
+            .first()
+
+        if update_time:
+          formatted_time = update_time.timestamp.strftime('%Y-%m-%d %H:%M:%S')
+          return formatted_time
+        else:
+         return None
+
+ 
+    def get_calculated_value(self, service_id: int, scorecard_id: int):
+     scorecard_metrics = self.scorecardMetrics.get_metric(scorecard_id)
+     print("Scorecard metrics:", scorecard_metrics)
+     metric_info_dict = {
+        metric.metricId: (metric.criteria, metric.desiredValue, metric.weight)
+        for metric in scorecard_metrics
+      }
+
+     metric_info_dict = {}
+     for metric in scorecard_metrics:
+        metric_type = self.Metric.getMetricType(metric.metricId) 
```

Comment:
query the DB once and get list of metrics corresponding to the list of IDs embedded in the scorecard_metrics. then extract the type information of each metric given its ID from the returned list.